analyzers/src/SonarAnalyzer.CSharp/Extensions/AssignmentExpressionSyntaxExtensions.cs

```diff
@@ -26,8 +26,114 @@
 {
-    public static class AssignmentExpressionSyntaxExtensions
+    internal static class AssignmentExpressionSyntaxExtensions
     {
+        public readonly record struct AssignmentMapping(CSharpSyntaxNode Left, ExpressionSyntax Right);
+
+        /// <summary>
+        /// Maps the left and the right side arguments of an <paramref name="assignment"/>. If both sides are tuples, the tuple elements are mapped.
+        /// <code>
+        /// (var x, var y) = (1, 2);                 // [x→1, y→2]
+        /// var (x, y) = (1, 2);                     // [x→1, y→2]
+        /// (var x, (var y, var z)) = (1, (2, 3));   // [x→1, y→2, z→3]
+        /// var x = 1;                               // [x→1]
+        /// (var x, var y) = M();                    // [(x,y)→M()]
+        /// </code>
+        /// </summary>
+        /// <param name="assignment">The <paramref name="assignment"/> expression.</param>
+        /// <returns>A mapping from expressions on the left side of the <paramref name="assignment"/> to the right side.</returns>
+        public static ImmutableArray<AssignmentMapping> MapAssignmentArguments(this AssignmentExpressionSyntax assignment)
+        {
+            // (var x, var y) = (1, 2)
+            if (TupleExpressionSyntaxWrapper.IsInstance(assignment.Left)
+                && TupleExpressionSyntaxWrapper.IsInstance(assignment.Right))
+            {
+                var left = (TupleExpressionSyntaxWrapper)assignment.Left;
+                var right = (TupleExpressionSyntaxWrapper)assignment.Right;
+                var arrayBuilder = ImmutableArray.CreateBuilder<AssignmentMapping>(left.Arguments.Count);
+                if (MapTupleElements(arrayBuilder, left, right))
+                {
+                    return arrayBuilder.ToImmutableArray();
+                }
+            }
+            // var (x, y) = (1, 2)
+            else if (DeclarationExpressionSyntaxWrapper.IsInstance(assignment.Left)
+                     && (DeclarationExpressionSyntaxWrapper)assignment.Left is { Designation: { } leftDesignation }
+                     && ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(leftDesignation)
+                     && TupleExpressionSyntaxWrapper.IsInstance(assignment.Right))
+            {
+                var left = (ParenthesizedVariableDesignationSyntaxWrapper)leftDesignation;
+                var right = (TupleExpressionSyntaxWrapper)assignment.Right;
+                var arrayBuilder = ImmutableArray.CreateBuilder<AssignmentMapping>(left.Variables.Count);
+                if (MapDesignationElements(arrayBuilder, left, right))
+                {
+                    return arrayBuilder.ToImmutableArray();
+                }
+            }
+
+            return ImmutableArray.Create(new AssignmentMapping(assignment.Left, assignment.Right));
+
+            static bool MapTupleElements(ImmutableArray<AssignmentMapping>.Builder arrayBuilder, TupleExpressionSyntaxWrapper left, TupleExpressionSyntaxWrapper right)
```

Comment:
These two local methods are static, quite big and cannot benefit from not being static => let's move them out as private static methods.

Reply:
The methods are part of AssignmentExpressionSyntaxExtensions, which might grow in the future. By making them local functions, the helper methods are kept in scope. I extracted them now and created a `#region'  to help with navigation.

Reply:
We're trying to avoid `#regions` unless there's a strong reason. Just move them to the end of the file as private. Optionally, you can swap `MapAssignmentArguments` and `AssignmentTargets` to keep them together.

Reply:
> Optionally, you can swap MapAssignmentArguments and AssignmentTargets to keep them together.

Or make them local functions. It scopes the helpers and keeps them together. From what I can tell, extension methods are grouped by node type in a common class but are often unrelated within the class. local functions are a good way to keep the helpers close to their callers. I would really like to make them static local functions. What do you think?